file_transfers.c

```diff
@@ -509,94 +482,123 @@ static void incoming_file_avatar(Tox *tox, uint32_t friend_number, uint32_t file
 }
 
 static void incoming_file_existing(Tox *tox, uint32_t friend_number, uint32_t file_number, uint32_t kind, uint64_t size, const uint8_t *filename, size_t filename_length, void *UNUSED(user_data)){
-    //new incoming file
     debug("FileTransfer:\tIncoming Existing file from friend (%u) \n", friend_number);
 
-    uint8_t new_id[TOX_FILE_ID_LENGTH] = {0};
-    tox_file_get_file_id(tox, friend_number, file_number, new_id, 0);
-    FILE_TRANSFER *file_new = get_file_transfer(friend_number, file_number);
-    FILE_TRANSFER *file_existing = utox_file_find_existing(new_id);
-
-    FILE *file;
-    TOX_ERR_FILE_SEEK error = 0;
-
-    if (!file_new) {
+    FILE_TRANSFER *file_handle = get_file_transfer(friend_number, file_number);
+    if (!file_handle) {
         tox_file_control(tox, friend_number, file_number, TOX_FILE_CONTROL_CANCEL, 0);
         return;
     }
 
-    if(file_existing){
-        uint8_t old_broken_number = file_existing->resume;
-        if( file_existing->friend_number == friend_number &&
-            file_existing->size == size &&
-            memcmp(file_existing->name, filename, filename_length) == 0 ){
-            // DO STUFF
-
-            utox_build_file_transfer(file_new, friend_number, file_number, size, 1, 0, 0, TOX_FILE_KIND_EXISTING,
-                filename, filename_length, NULL, 0, new_id, tox);
-
-            file_new->size_transferred = file_existing->size_transferred;
-            file_new->file = file_existing->file;
-            file_new->ui_data = file_existing->ui_data;
-            utox_file_free_resume(old_broken_number);
-            file_new->resume = utox_file_alloc_resume(tox, file_new);
+    uint8_t file_id[TOX_FILE_ID_LENGTH];
+    tox_file_get_file_id(tox, friend_number, file_number, file_id, 0);
+    if(memcmp(file_handle->file_id, file_id, TOX_FILE_ID_LENGTH) == 0){
+        if(file_handle->size != size){
+            debug("FileTransfer:\tResume size mismatch, aborting!\n");
+            file_transfer_local_control(tox, friend_number, file_number, TOX_FILE_CONTROL_CANCEL);
+            return;
+        } else {
+            debug("FileTransfer:\tIncoming existing file approved!\n");
+            tox_file_seek(tox, friend_number, file_number, file_handle->size_transferred, 0);
+            file_transfer_local_control(tox, friend_number, file_number, TOX_FILE_CONTROL_RESUME);
+            return;
+        }
+    } else {
+        /* File number wasn't the same, maybe it's still somewhere? */
+        int good = 0; /* Used to skip searching for files */
+        FILE_TRANSFER *file_search;
+        for(int i = 0 ; i < MAX_FILE_TRANSFERS ; i++){
+            file_search = get_file_transfer(friend_number, i);
+            if(memcmp(file_search->file_id, file_id, TOX_FILE_ID_LENGTH) == 0){
+                memcpy(file_handle, file_search, sizeof(*file_handle));
+                file_handle->file_number = file_number;
```

Comment:
Dirty hack to find out if uTox was able to find that file in memory with a
different file number

On Wed, Apr 22, 2015, 13:30 irungentoo <notifications@github.com> wrote:

> In file_transfers.c
> <https://github.com/notsecure/uTox/pull/972#discussion_r28911677>:
>
> > +            return;
> > +        } else {
> > +            debug("FileTransfer:\tIncoming existing file approved!\n");
> > +            tox_file_seek(tox, friend_number, file_number, file_handle->size_transferred, 0);
> > +            file_transfer_local_control(tox, friend_number, file_number, TOX_FILE_CONTROL_RESUME);
> > +            return;
> > +        }
> > +    } else {
> > +        /* File number wasn't the same, maybe it's still somewhere? */
> > +        int good = 0; /* Used to skip searching for files */
> > +        FILE_TRANSFER *file_search;
> > +        for(int i = 0 ; i < MAX_FILE_TRANSFERS ; i++){
> > +            file_search = get_file_transfer(friend_number, i);
> > +            if(memcmp(file_search->file_id, file_id, TOX_FILE_ID_LENGTH) == 0){
> > +                memcpy(file_handle, file_search, sizeof(*file_handle));
> > +                file_handle->file_number = file_number;
>
> good = 1;
>
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/notsecure/uTox/pull/972/files#r28911677>.
>